File: KunalKushwaha/Assignments/Tree/FindModeInBinarySearchTree.java
package KunalKushwaha.Assignments.Tree;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Queue;

public class FindModeInBinarySearchTree {
	public int[] findMode(TreeNode root) {
		Map<Integer, Integer> map = new HashMap<>();
		Queue<TreeNode> queue = new LinkedList<>();
		queue.offer(root);
		int max = Integer.MIN_VALUE;

		while (!queue.isEmpty()) {
			int n = queue.size();
			for (int i = 0; i < n; i++) {
				TreeNode curr = queue.poll();
				map.put(curr.val, map.getOrDefault(curr.val, 0) + 1);
				max = Math.max(max, map.get(curr.val));
				if (curr.left != null)
					queue.offer(curr.left);
				if (curr.right != null)
					queue.offer(curr.right);
			}
		}

		List<Integer> modes = new ArrayList<>();
		for (Integer key : map.keySet()) {
			if (map.get(key) == max) {
				modes.add(key);
			}
		}

		return modes.stream().mapToInt(Integer::intValue).toArray();
	}
}

File: KunalKushwaha/Assignments/Tree/FindModeInBinarySearchTree.py
from collections import deque
from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
	def findMode(self, root: Optional[TreeNode]) -> List[int]:
		counts = {}
		max_count = 0
		modes = []
		def inorder(node):
			if not node:
				return
			inorder(node.left)
			nonlocal max_count, modes
			counts[node.val] = 1 + counts.get(node.val, 0)
			
			if counts[node.val] > max_count:
				max_count = counts[node.val]
				modes = [node.val]
			elif counts[node.val] == max_count:
				modes.append(node.val)
			inorder(node.right)
		inorder(root)
		return modes